source/common/stats/thread_local_store.cc

```diff
@@ -206,11 +206,13 @@ void ThreadLocalStoreImpl::mergeHistograms(PostMergeCb merge_complete_cb) {
     ASSERT(!merge_in_progress_);
     merge_in_progress_ = true;
     tls_->runOnAllThreads(
-        [this]() -> void {
-          for (const auto& id_hist : tls_->getTyped<TlsCache>().tls_histogram_cache_) {
+        [](ThreadLocal::ThreadLocalObjectSharedPtr object)
```

Comment:
A thought for a future revision, I think. For what I think is maybe nicer compile-time type safety, as well as less cumbersome usage, can we consider 2 changes:

1. Templatize the usage of slotes, so I put the fact that this is slotting a `TlsCache` into my declaration: `ThreadLocal::SlotPtr<TlsCache>` and then, for me, asType() is not needed.
2. Pass in object here as a non-const reference, so I can update if I like, but don't have to return it or declare the return-value.

Then this call-site becomes:

```
tls_->runOnAllThreads([](TlsCache& tls_cache) {
  for (const auto& id_hist : tls_cache.tls_histogram_cache_) {
    const TlsHistogramSharedPtr& tls_hist = id_hist.second;
    tls_hist->beginMerge();
  }
});
```

I think the Slot mechanism itself doesn't to change at all to accommodate this templating layer. It can do what it does now, and the templating API can be a thin header-only layer above.

Reply:
Let me think about this more. I thought about the type of templating you suggest, but I think I decided it would get complicated with the APIs/callbacks. I can try to prototype it. I agree with the idea around passing a non-cost ref as a way of simplifying it. I'm going to open a follow up issue about adding clang-tidy and will add this also.

Reply:
https://github.com/envoyproxy/envoy/issues/13313

Reply:
awesome, thanks! I'll follow along in that bug.